elyra/pipeline/validation.py

```diff
@@ -165,17 +165,32 @@ def _validate_pipeline_structure(self, pipeline: dict, response: ValidationRespo
         elif 'version' not in primary_pipeline['app_data']:
             response.add_message(severity=ValidationSeverity.Error, message_type="invalidPipeline",
                                  message="Pipeline app_data 'version' is missing from primary pipeline")
-        elif not int(primary_pipeline['app_data']['version']) > 0:
+
+        # validate pipeline version compatibility
+        pipeline_version = int(primary_pipeline['app_data']['version'])
+        if pipeline_version not in range(PIPELINE_CURRENT_VERSION + 1):
             response.add_message(severity=ValidationSeverity.Error, message_type="invalidPipeline",
                                  message="Primary pipeline version field has an invalid value")
 
+        elif pipeline_version < PIPELINE_CURRENT_VERSION:
+            # Pipeline needs to be migrated
+            response.add_message(severity=ValidationSeverity.Error, message_type="invalidPipeline",
+                                 message=f'Pipeline version {pipeline_version} is out of date and needs to be migrated '
+                                         f'using the Elyra pipeline editor.')
+        elif pipeline_version > PIPELINE_CURRENT_VERSION:
+            # New version of Elyra is needed
+            response.add_message(severity=ValidationSeverity.Error, message_type="invalidPipeline",
+                                 message='Pipeline was last edited in a newer version of Elyra. '
+                                         'Update Elyra to use this pipeline.')
+
+        # Validate pipeline schema version
         if 'version' not in pipeline_json:
             response.add_message(severity=ValidationSeverity.Error, message_type="invalidPipeline",
                                  message="Primary pipeline version field is missing")
         if not isinstance(pipeline_json['version'], str):
             response.add_message(severity=ValidationSeverity.Error, message_type="invalidPipeline",
                                  message="Primary pipeline version field should be a string.")
-        if float(pipeline_json['version']) < current_pipeline_schema_version:
+        if float(pipeline_json['version']) < PIPELINE_CURRENT_SCHEMA:
```

Comment:
Should version identifiers other than PIPELINE_CURRENT_SCHEMA (e.g. > PIPELINE_CURRENT_SCHEMA) be supported at all? As is,  a newer schema version would not trigger an error, yet nobody knows atm what changes PIPELINE_CURRENT_SCHEMA might introduce.

Reply:
In line below see earlier comment, about providing contextual information that the caller might want to expose

```
...,
data = {
  supported_schema_version: PIPELINE_CURRENT_SCHEMA,
  detected_schema_version: ...
}
```